src/components/transition/Slider.tsx

```diff
@@ -0,0 +1,95 @@
+
+  // Whether the content is currently partially or wholly visible. This is
+  // different from `visible` when collapsing as it is true until the collapse
+  // animation completes.
+  const [contentVisible, setContentVisible] = useState(visible);
+
+  // Adjust the container height when the `visible` prop changes.
+  useEffect(() => {
+    const isVisible = containerHeight !== 0;
+    if (visible === isVisible) {
+      // Do nothing after the initial mount.
+      return;
+    }
+
+    const el = containerRef.current!;
+    if (visible) {
+      // Show the content synchronously so that we can measure it here.
+      el.style.display = '';
+
+      // Make content visible in future renders.
+      setContentVisible(true);
+
+      // When expanding, transition the container to the current fixed height
+      // of the content. After the transition completes, we'll reset to "auto"
+      // height to adapt to future content changes.
+      setContainerHeight(el.scrollHeight);
+    } else {
+      // When collapsing, immediately change the current height to a fixed height
+      // (in case it is currently "auto"), force a synchronous layout,
+      // then transition to 0.
+      //
+      // These steps are needed because browsers will not animate transitions
+      // from "auto" => "0" and may not animate "auto" => fixed height => 0
+      // if the layout tree transitions directly from "auto" => 0.
+      el.style.height = `${el.scrollHeight}px`;
+
+      // Force a sync layout.
+      el.getBoundingClientRect();
+
+      setContainerHeight(0);
+    }
+  }, [containerHeight, visible]);
+
+  const handleTransitionEnd = useCallback(() => {
+    if (visible) {
+      setContainerHeight('auto');
+      onTransitionEnd?.('in');
+    } else {
+      // When the collapse animation completes, stop rendering the content so
+      // that the browser has fewer nodes to render and the content is removed
+      // from keyboard navigation.
+      setContentVisible(false);
+      onTransitionEnd?.('out');
+    }
+  }, [setContainerHeight, visible, onTransitionEnd]);
+
+  const isFullyVisible = containerHeight === 'auto';
+
+  return (
+    <div
+      // nb. Preact uses "ontransitionend" rather than "onTransitionEnd".
+      // See https://bugs.chromium.org/p/chromium/issues/detail?id=961193
+      //
+      // @ts-ignore
+      // eslint-disable-next-line react/no-unknown-property
+      ontransitionend={handleTransitionEnd}
+      ref={containerRef}
+      style={{
+        display: contentVisible ? '' : 'none',
+        height: containerHeight,
+
+        // When the Slider is fully open, overflow is made visible so that
+        // focus rings, which may extend outside the bounds of the Slider content,
+        // are visible.
+        overflow: isFullyVisible ? 'visible' : 'hidden',
+        transition: `height 0.15s ease-in`,
+      }}
+    >
+      {children}
+    </div>
+  );
+};
```

Comment:
This component is brought from [client](https://github.com/hypothesis/client/blob/6f3a741d630375c9dfd780af3a66991c92027c7e/src/sidebar/components/Slider.tsx) verbatim.

Reply:
Pro tip! GitHub now has the ability to add a comment at a module level, if that's helpful (the comment box icon at the top of each file's diff):

<img width="253" alt="image" src="https://user-images.githubusercontent.com/439947/233651135-e447bf7c-6d41-47c9-beb5-db90ef7042f1.png">

Reply:
I actually knew that! 😅